### PR TITLE
Cache entity variable existence from SQL database

### PR DESCRIPTION
**Pull request by shifucun**
* Query all <entity, variable> pair from custom DC SQL database and cache them in-memory
* Improve the count function by using this cache information.

### CODE REVIEW

File: internal/sqldb/sqlquery/entity_variable.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	"database/sql"
+	"time"
+
+	"github.com/datacommonsorg/mixer/internal/util"
+)
+
+// EntityVariable returns all existent entity, variable pairs
+func EntityVariable(sqlClient *sql.DB) (map[string]map[string]struct{}, error) {
+	defer util.TimeTrack(time.Now(), "SQL: EntityVariable")
+	query := "SELECT DISTINCT entity, variable FROM observations o"
```
Comment (keyurva) [inline internal/sqldb/sqlquery/entity_variable.go:27]: A couple questions:

1. How long does this query take on Cloud SQL for the large ONE dataset?
2. Any concerns about the size of this map? If there are 200 entities and 6000 vars, there will be 1.2M entries.

File: internal/server/cache/cache.go [hunk 2]
```diff
@@ -49,8 +49,10 @@ type Cache struct {
 	blocklistSvgs map[string]struct{}
 	// SVG search index
 	svgSearchIndex *resource.SearchIndex
-	// Custom provenance from SQL storage
-	customProvenances map[string]*pb.Facet
+	// Provenance from SQL storage
+	sqlProvenances map[string]*pb.Facet
+	// SQL database entity, variable existence pairs
+	sqlEntityVariable map[string]map[string]struct{}
```
Comment (keyurva) [inline internal/server/cache/cache.go:55]: This article talks about using struct keys in such situations. Consider using it if it makes things easier. https://go.dev/blog/maps#key-types

File: internal/server/cache/cache.go [hunk 3]
```diff
@@ -117,13 +123,17 @@ func NewCache(
...
+		if err != nil {
+			return nil, err
+		}
+		c.sqlProvenances = sqlProv
+		sqlExistenceMap, err := sqlquery.EntityVariable(store.SQLClient)
 		if err != nil {
 			return nil, err
 		}
-		c.customProvenances = customProv
+		c.sqlEntityVariable = sqlExistenceMap
```
Comment (keyurva) [inline internal/server/cache/cache.go:136]: Actually `sqlExistenceMap` might be a good name for the cache entry itself.

File: internal/server/cache/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shifucun) [reply internal/server/cache/cache.go]: Changed to sqlExistenceMap

Comment (shifucun) [reply internal/server/cache/cache.go]: This is a very good ideal. Updated

File: internal/server/cache/cache.go (hunk 3 above)
Comment (shifucun) [reply internal/server/cache/cache.go]: Done

File: internal/sqldb/sqlquery/entity_variable.go (hunk 1 above)
Comment (shifucun) [reply internal/sqldb/sqlquery/entity_variable.go]: 1. 6-7 seconds
2. Give the map is keyed by <e,v> with value of an empty struct, each entry takes O(10) bytes, so for this case it's in the order 12Mb (with a factor of 5, it's 60Mb) which is a reasonable size.

**Pull request merged**